File: backend/app/schemas/import_job.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.dicom_import import ImportReport


class ImportJobStatus(StrEnum):
    UPLOADING = "uploading"
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class ImportManifestFile(BaseModel):
    model_config = ConfigDict(extra="forbid")

    relative_path: str = Field(min_length=1, max_length=1024)
    size_bytes: int = Field(ge=1, le=512 * 1024 * 1024)
    last_modified_ms: int = Field(ge=0)
    resume_fingerprint: str = Field(pattern=r"^[0-9a-f]{64}$")


class ImportJobCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    files: list[ImportManifestFile] = Field(min_length=1, max_length=2000)


class ImportJobFileRead(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    ordinal: int = Field(ge=0, le=1999)
    relative_path: str = Field(min_length=1, max_length=1024)
    size_bytes: int = Field(ge=1, le=512 * 1024 * 1024)
    last_modified_ms: int = Field(ge=0)
    resume_fingerprint: str = Field(pattern=r"^[0-9a-f]{64}$")
    confirmed_offset: int = Field(ge=0)


class ImportJobRead(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    patient_id: UUID
    status: ImportJobStatus
    total_files: int = Field(ge=1, le=2000)
    total_bytes: int = Field(ge=1, le=8 * 1024**3)
    uploaded_bytes: int = Field(ge=0)
    files: list[ImportJobFileRead] = Field(min_length=1, max_length=2000)
    report: ImportReport | None = None
    error_code: str | None = Field(default=None, max_length=64)
    error_message: str | None = Field(default=None, max_length=512)
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None
    completed_at: datetime | None

    @field_validator("created_at", "updated_at", "started_at", "completed_at", mode="before")
    @classmethod
    def restore_utc_timezone(cls, value: datetime | None) -> datetime | None:
        if value is None:
            return None
        if value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)
        return value.astimezone(timezone.utc)


class ImportUploadProgressRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    file_id: UUID
    confirmed_offset: int = Field(ge=0)
    uploaded_bytes: int = Field(ge=0)
    total_bytes: int = Field(ge=1, le=8 * 1024**3)

File: frontend/src/features/dicom-import/model/importJob.ts
import type { ImportReport } from './dicomImport'


export type ImportJobStatus =
  | 'uploading'
  | 'queued'
  | 'running'
  | 'completed'
  | 'failed'

export interface ImportManifestFile {
  relative_path: string
  size_bytes: number
  last_modified_ms: number
  resume_fingerprint: string
}

export interface ImportJobFile extends ImportManifestFile {
  id: string
  ordinal: number
  confirmed_offset: number
}

export interface ImportJob {
  id: string
  patient_id: string
  status: ImportJobStatus
  total_files: number
  total_bytes: number
  uploaded_bytes: number
  files: ImportJobFile[]
  report: ImportReport | null
  error_code: string | null
  error_message: string | null
  created_at: string
  updated_at: string
  started_at: string | null
  completed_at: string | null
}

export interface ImportUploadProgress {
  file_id: string
  confirmed_offset: number
  uploaded_bytes: number
  total_bytes: number
}

export interface UploadProgress {
  uploadedBytes: number
  totalBytes: number
  currentFile: number
  totalFiles: number
}
